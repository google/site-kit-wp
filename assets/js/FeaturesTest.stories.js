import { useSelect } from 'googlesitekit-data';
import { WithTestRegistry } from '../../tests/js/utils';
import { CORE_SITE } from './googlesitekit/datastore/site/constants';
import { useFeature } from './hooks/useFeature';

function FeaturesTest() {
	const isFooEnabled = useFeature( 'foo' );

	return <p>Feature foo is { isFooEnabled ? 'enabled' : 'disabled' }</p>;
}

function WasFeatureEnabledInCallback() {
	const wasFooEnabled = useSelect( ( select ) =>
		select( CORE_SITE ).wasFeatureEnabled( 'foo' )
	);

	return (
		<p>
			Was feature foo enabled during callback?&nbsp;
			{ wasFooEnabled ? 'yes' : 'no' }
		</p>
	);
}

export function WithDefaultDecorator() {
	return (
		<div>
			<h1>With Default Decorator</h1>
			<FeaturesTest />
		</div>
	);
}
WithDefaultDecorator.parameters = {
	features: [ 'foo' ],
};

export function WithTestRegistryInStory() {
	return (
		<div>
			<h1>With Test Registry In Story</h1>
			<WithTestRegistry
				features={ [ 'foo' ] }
				callback={ ( registry ) => {
					registry.dispatch( CORE_SITE ).checkFeatureEnabled( 'foo' );
				} }
			>
				<FeaturesTest />
				<WasFeatureEnabledInCallback />
			</WithTestRegistry>
		</div>
	);
}

export default {
	title: 'FeaturesTest',
	component: FeaturesTest,
};
