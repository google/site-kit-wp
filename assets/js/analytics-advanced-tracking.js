/**
 * Analytics advanced tracking script to be inserted into the frontend via PHP.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// This file should not use any dependencies because it is used in the frontend.

/**
 * Internal dependencies
 */
import setUpAdvancedTracking from './analytics-advanced-tracking/set-up-advanced-tracking';

/**
 * Sends a tracking event to Analytics via gtag.
 *
 * @since 1.18.0
 *
 * @param {string} action   Event action / event name.
 * @param {Object} metadata Additional event metadata to send, or `null`.
 */
function sendEvent( action, metadata ) {
	window.gtag( 'event', action, metadata || undefined ); // eslint-disable-line no-restricted-globals
}

const events = window._googlesitekitAnalyticsTrackingData || []; // eslint-disable-line no-restricted-globals
if ( Array.isArray( events ) ) {
	setUpAdvancedTracking( events, sendEvent );
	// eslint-disable-next-line no-console
	console.log( 'Hello QA testing' );
}
