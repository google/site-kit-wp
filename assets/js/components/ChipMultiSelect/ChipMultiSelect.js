/**
 * ChipMultiSelect component.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import { Provider } from './ChipMultiSelectContext';

export default function ChipMultiSelect( {
	children,
	onToggleChip = () => {},
} ) {
	return (
		<Provider value={ { onToggleChip } }>
			<div className="googlesitekit-chip-multi-select">{ children }</div>
		</Provider>
	);
}

ChipMultiSelect.propTypes = {
	children: PropTypes.node.isRequired,
	onToggleChip: PropTypes.func,
};
