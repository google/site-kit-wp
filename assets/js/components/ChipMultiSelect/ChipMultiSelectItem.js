/**
 * ChipMultiSelectItem component.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { omit } from 'lodash';
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { useContext } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { Chip } from 'googlesitekit-components';
import ChipMultiSelectContext from './ChipMultiSelectContext';
import { ChipCheckmark } from './ChipCheckmark';
export default function ChipMultiSelectItem( {
	children,
	id,
	selected,
	...props
} ) {
	const { onToggleChip } = useContext( ChipMultiSelectContext );

	return (
		<Chip
			className="googlesitekit-chip-multi-select__item"
			CheckMark={ ChipCheckmark }
			onClick={ () => {
				// Indicate the desired state of the chip.
				onToggleChip( id, ! selected );
			} }
			id={ id }
			label={ children }
			selected={ selected }
			{ ...props }
		/>
	);
}

ChipMultiSelectItem.propTypes = {
	children: PropTypes.string.isRequired,
	id: PropTypes.string.isRequired,
	...omit( Chip.propTypes, 'label' ),
};
