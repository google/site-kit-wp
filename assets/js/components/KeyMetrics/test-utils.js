/**
 * Key Metrics test utility functions.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { CORE_WIDGETS } from '../../googlesitekit/widgets/datastore/constants';
import { AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY } from '../../googlesitekit/widgets/default-areas';
import { CONTEXT_MAIN_DASHBOARD_KEY_METRICS } from '../../googlesitekit/widgets/default-contexts';

/**
 * Provides key metric widgets registration data to the given registry.
 *
 * This is a temporary utility until a solidified solution is implemented to effortlessly provide
 * widget registrations to the registry.
 *
 * @since 1.105.0
 * @see {@link https://github.com/google/site-kit-wp/issues/7264} Initiative to implement a utility to provide widget registrations.
 *
 * @param {Object} registry The registry to set up.
 * @param {Object} widgets  Object containing options mapped to widget slugs.
 */
export function provideKeyMetricsWidgetRegistrations( registry, widgets ) {
	registry
		.dispatch( CORE_WIDGETS )
		.registerWidgetArea( AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY, {
			title: 'Key metrics',
		} );

	registry
		.dispatch( CORE_WIDGETS )
		.assignWidgetArea(
			AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY,
			CONTEXT_MAIN_DASHBOARD_KEY_METRICS
		);

	Object.keys( widgets ).forEach( ( slug ) => {
		registry.dispatch( CORE_WIDGETS ).registerWidget( slug, {
			Component() {
				return <div>Hello test.</div>;
			},
			...widgets[ slug ],
		} );

		registry
			.dispatch( CORE_WIDGETS )
			.assignWidget( slug, AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY );
	} );
}
