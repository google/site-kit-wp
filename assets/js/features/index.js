/**
 * Feature flags.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const enabledFeatures = new Set(
	global?._googlesitekitBaseData?.enabledFeatures || []
);

/**
 * Returns true if a feature is enabled; false otherwise.
 *
 * @since 1.25.0
 * @since 1.33.0 Changed _enabledFeatures argument to be a `Set` instead of `Array`.
 *
 * @param {string} feature            The name of the feature to check.
 * @param {Set}    [_enabledFeatures] Optional. The set of enabled features. Uses `enabledFeatures` set by the server in a global JS variable, by default.
 * @return {boolean} `true` if a feature is enabled; `false` otherwise.
 */
export function isFeatureEnabled(
	feature,
	_enabledFeatures = enabledFeatures
) {
	if ( ! ( _enabledFeatures instanceof Set ) ) {
		return false;
	}
	return _enabledFeatures.has( feature );
}
