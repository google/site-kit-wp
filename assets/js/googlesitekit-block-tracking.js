import { watchBlocks } from './googlesitekit/block-tracking/watch-blocks';

const blocksToTrack = [
	'google-site-kit/rrm-subscribe-with-google',
	'google-site-kit/rrm-contribute-with-google',
];

watchBlocks( blocksToTrack );
