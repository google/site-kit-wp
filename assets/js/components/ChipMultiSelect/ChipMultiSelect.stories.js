/**
 * ChipMultiSelect Component Stories.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { useState } from '@wordpress/element';

/**
 * Internal dependencies
 */
import ChipMultiSelect from './ChipMultiSelect';
import ChipMultiSelectItem from './ChipMultiSelectItem';

function Template( { initialSelectedChips = {} } ) {
	const [ selectedChips, setSelectedChips ] =
		useState( initialSelectedChips );

	return (
		<ChipMultiSelect
			onToggleChip={ ( id, isSelected ) => {
				global.console.log( 'onToggleChip', id, isSelected );

				const newSelectedChips = selectedChips[ id ]
					? { ...selectedChips, [ id ]: ! selectedChips[ id ] }
					: { ...selectedChips, [ id ]: true };

				setSelectedChips( newSelectedChips );
			} }
		>
			<ChipMultiSelectItem id="posts" selected={ selectedChips.posts }>
				Posts
			</ChipMultiSelectItem>
			<ChipMultiSelectItem id="pages" selected={ selectedChips.pages }>
				Pages
			</ChipMultiSelectItem>
			<ChipMultiSelectItem
				id="customPostType"
				selected={ selectedChips.customPostType }
			>
				Custom Post Type
			</ChipMultiSelectItem>
		</ChipMultiSelect>
	);
}

export const Default = Template.bind( {} );
Default.storyName = 'Default';
Default.scenario = {};

export const WithSelectedChips = Template.bind( {} );
WithSelectedChips.storyName = 'With Selected Chips';
WithSelectedChips.args = {
	initialSelectedChips: {
		posts: true,
		pages: false,
		customPostType: true,
	},
};
WithSelectedChips.scenario = {};

export default {
	title: 'Components/ChipMultiSelect',
	component: ChipMultiSelect,
	decorators: [
		( Story ) => (
			<div className="googlesitekit-layout" style={ { padding: '24px' } }>
				<Story />
			</div>
		),
	],
};
