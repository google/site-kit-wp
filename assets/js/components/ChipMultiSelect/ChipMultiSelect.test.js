/**
 * ChipMultiSelect component tests.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { render, fireEvent } from '../../../../tests/js/test-utils';
import ChipMultiSelect from './ChipMultiSelect';
import ChipMultiSelectItem from './ChipMultiSelectItem';

describe( 'ChipMultiSelect', () => {
	it( 'renders correctly', () => {
		const { container } = render(
			<ChipMultiSelect>
				<ChipMultiSelectItem id="channels">
					Channels
				</ChipMultiSelectItem>
				<ChipMultiSelectItem id="locations" selected>
					Locations
				</ChipMultiSelectItem>
				<ChipMultiSelectItem id="devices">Devices</ChipMultiSelectItem>
			</ChipMultiSelect>
		);

		expect( container ).toMatchSnapshot();
	} );

	it( 'calls `onToggleChip()` when a chip is selected', () => {
		const onToggleChip = jest.fn();

		const { getByText } = render(
			<ChipMultiSelect onToggleChip={ onToggleChip }>
				<ChipMultiSelectItem id="channels">
					Channels
				</ChipMultiSelectItem>
				<ChipMultiSelectItem id="locations" selected>
					Locations
				</ChipMultiSelectItem>
				<ChipMultiSelectItem id="devices">Devices</ChipMultiSelectItem>
			</ChipMultiSelect>
		);

		fireEvent.click( getByText( 'Channels' ) );

		expect( onToggleChip ).toHaveBeenCalledWith( 'channels', true );
	} );

	it( 'calls `onToggleChip()` when a chip is unselected', () => {
		const onToggleChip = jest.fn();

		const { getByText } = render(
			<ChipMultiSelect onToggleChip={ onToggleChip }>
				<ChipMultiSelectItem id="channels">
					Channels
				</ChipMultiSelectItem>
				<ChipMultiSelectItem id="locations" selected>
					Locations
				</ChipMultiSelectItem>
				<ChipMultiSelectItem id="devices">Devices</ChipMultiSelectItem>
			</ChipMultiSelect>
		);

		fireEvent.click( getByText( 'Locations' ) );

		expect( onToggleChip ).toHaveBeenCalledWith( 'locations', false );
	} );
} );
