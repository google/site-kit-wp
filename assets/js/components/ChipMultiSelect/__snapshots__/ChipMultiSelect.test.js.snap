// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChipMultiSelect renders correctly 1`] = `
<div>
  <div
    class="googlesitekit-chip-multi-select"
  >
    <div
      class="mdc-chip googlesitekit-chip googlesitekit-chip-multi-select__item"
      data-chip-id="channels"
      tabindex="0"
    >
      <div
        class="mdc-chip__checkmark"
      >
        <svg />
      </div>
      <div
        class="mdc-chip__text"
      >
        Channels
      </div>
    </div>
    <div
      class="mdc-chip mdc-chip--selected googlesitekit-chip googlesitekit-chip-multi-select__item"
      data-chip-id="locations"
      tabindex="0"
    >
      <div
        class="mdc-chip__checkmark"
      >
        <svg />
      </div>
      <div
        class="mdc-chip__text"
      >
        Locations
      </div>
    </div>
    <div
      class="mdc-chip googlesitekit-chip googlesitekit-chip-multi-select__item"
      data-chip-id="devices"
      tabindex="0"
    >
      <div
        class="mdc-chip__checkmark"
      >
        <svg />
      </div>
      <div
        class="mdc-chip__text"
      >
        Devices
      </div>
    </div>
  </div>
</div>
`;
